README
 liboping – Library to ping IPv4 and IPv6 hosts in parallel
════════════════════════════════════════════════════════════
http://noping.cc/

About
━━━━━

  liboping was inspired by ping, libping and fping: It differs from these
  existing solutions in that it can “ping” multiple hosts in parallel using
  IPv4 or IPv6 transparently. Other design principles were an object oriented
  interface, simplicity and extensibility.

  On top of liboping two command line applications have been built. “oping” is
  a drop-in replacement for ping(1) with very similar output. “noping” is an
  ncurses-based application which displays statistics while pinging and
  highlights aberrant round-trip times.


Features
━━━━━━━━

  • Support for multiple hosts.

  • Support for IPv4 and IPv6.

  • Object oriented interface.


Perl bindings
━━━━━━━━━━━━━

  Included in the source package of liboping are bindings for Perl. The code
  resides in the bindings/ subdirectory and is compiled and installed by
  default. To disable building the Perl bindings, call “configure” with
  “--without-perl-bindings”.


Permissions
━━━━━━━━━━━━━

  On UNIX, special permissions are required to open raw sockets (raw(7)). If
  you compile and install the “oping” and “noping” binaries as normal user
  (which is strongly suggested), you won't be able to use the binaries as a
  normal user, because you won't have the permission to open raw sockets.

  The “install” target will automatically try fix this, if it is run with UID~0
  (as user root).  When on Linux, the capabilities described below will be
  added. On other UNIXes the traditional Set-UID method (also described below)
  is used instead. The build system will not abort if this fails, because there
  are file systems which do not support either method. Also, the Debian
  packaging system and possibly other scenarios only act as if they were
  running as root.

  Linux
  ━━━━━
  On Linux, the preferred method is to assign the required “capability” to the
  binaries. This will allow the binary to open raw sockets, but doesn't give
  any other permissions such as reading other users' files or shutting down the
  system. The downside is that this mechanism is comparatively new: Assigning
  capabilities to files is available since Linux 2.6.24.

  To set the required capabilities, run (as user root):

    # setcap cap_net_raw=ep /opt/oping/bin/oping
    # setcap cap_net_raw=ep /opt/oping/bin/noping

  Other UNIX
  ━━━━━━━━━━
  Capabilities are a nice but Linux-specific solution. To make “oping” and
  “noping” available to unprivileged users on other UNIX systems, use the
  traditional set-UID root solution. If your system supports “saved set-UIDs”
  (basically all systems do), the applications will drop the privileges during
  initialization and only regain them when actually opening the socket(s).

  To set the set-UID bit, run (as user root):

    # chown root: /opt/oping/bin/{,n}oping
    # chmod u+s   /opt/oping/bin/{,n}oping


Licensing terms
━━━━━━━━━━━━━━━

  liboping is licensed under the “GNU Lesser General Public License” (LGPL),
  version 2.1 or later. The exact licensing terms can be found in the file
  “COPYING” included in the source distribution of liboping.

  The “oping” and “noping” utilities included in this package are licensed
  under the “GNU General Public License” (GPL), version 2. The full licensing
  terms can be found online at <http://www.gnu.org/licenses/gpl-2.0.html>.


Author
━━━━━━

  Florian “octo” Forster <ff at octo.it>



ChangeLog
2017-05-11, Version 1.10.0:
	* Build system: Search the "rt" library for clock_gettime(). This
	  fixes build issues on Mac OS X. Thanks to Yann E. Morni for this
	  fix. #9
	* Build system: The "--with[out]-ncurses" option has been added and
	  allows to enforce/disable the building of "noping". Thanks to Thomas
	  Deutschmann for the patch. #15
	* Build system: Compatibility code to work with pkg-config < 0.27 has
	  been added. #22
	* liboping: Creation of ICMPv4 packets has been fixed: due to an
	  incorrect sizeof() the payload may have been prefixed by 20 zero
	  bytes, resulting in larger packets than desired. Thanks to Kyle Zhou
	  for reporting and fixing this. #10
	* liboping: The number of file descriptors used has been reduced to at
	  most two (from one per host). This and a few other optimizations
	  significantly improve performance. Thanks to Luke Heberling for the
	  patch. #11
	* oping, noping: Handling of the "-O" command line flag has been
	  fixed.
	* oping, noping: The "-b" option has been added and enables a bell
	  whenever an echo reply is received. Thanks to Antoine Beaupré for
	  the patch. #6
	* noping: The background color has been changed to use the terminal
	  default. Thanks to @middleO. #18
	* noping: The ability to add hosts after noping has started (the "a"
	  key) has been added. Thanks to Hamish Coleman for the patch. #20,
	  #23

2016-06-27, Version 1.9.0:
	* liboping: The new "PING_OPT_MARK" option allows to mark packets,
	  which can be used for filtering and routing such packets on Linux.
	* oping, noping: The new "-m" command line option allows to set a mark
	  on packets sent by the tool.
	* oping, noping: The new "-O" command line option allows to write
	  measurements to an CSV file.
	* oping, noping: The new "-w" command line option allows to specify
	  the timeout after which a packet/reply is considered "dropped".

2014-11-20, Version 1.8.0:
	* oping, noping: Average and standard deviation have been removed from
	  the status output, which show median and 95th percentile instead.
	  The percentile can be chosen with the "-P" option.
	* noping: The additional graph types "histogram" and "boxplot" have
	  been added, which can be selected with the "-g" option.

2014-09-25, Version 1.7.0:
	* oping, noping: The new -Z option allows the exit status to indicate
	  the number of failing hosts. Thanks to Barak Pearlmutter for the
	  patch.
	* noping: The ability to print a "prettyping" style graph has been
	  added. Thanks to Antoine Beaupré for his work!
	* src/liboping.c: Build issues on Solaris have been fixed. Thanks
	  Scott Severtson for the fix!
	* Build system: Creation and installation of a pkg-config file has
	  been added. Thanks to Barak Pearlmutter for the patch.

2012-01-31, Version 1.6.2:
	* Build system: Setting capabilities and the set-UID bit has been made
	  more fault-tolerant, so that it will work with Debian's fakeroot(1)
	  utility.
	* src/liboping.c: Fixed a compiler warning about an non-static format
	  string. Thanks to Brian Edwards for pointing this out.
	* src/liboping.c: Fixed compilation under Mac OS X and Solaris. Thanks
	  to Clayton O'Neill for his patch.

2011-03-06, Version 1.6.1:
	* Build system: If "make install" is executed as root, the CAP_NET_RAW
	  capability is added to the binary (on Linux) or the set-UID bit is
	  set (other Unixes).
	* src/oping.c: Fix compiler warnings which may abort the build. Thanks
	  to James Bromberger for reporting the problem.
	* noping: Compatibility with ncurses 5.8 has been fixed. Thanks to
	  Gaetan Bisson for his patch.

2011-01-26, Version 1.6.0:
	* liboping: Improve timing of received network packets using
	  SO_TIMESTAMP if available. Thanks to Bruno Prémont for his patch.

2010-11-17, Version 1.5.1:
	* oping, noping: Alias for the “Voice Admit” DSCP has been added.
	* src/oping.c, src/liboping.c: Compiler warnings / errors have been
	  fixed. Thanks to James Bromberger for reporting one of them.

2010-10-27, Version 1.5.0:
	* src/liboping.c: The possibility to set the QoS byte of outgoing IPv4
	  and IPv6 packets and read the byte from incoming packets has been
	  added. Thanks to Vladimir Melnikov for his patch.
	* oping, noping: Add the ability to configure the QoS field on the
	  command line. If either the QoS field of outgoing or incoming
	  packets is non-standard, the QoS byte of incoming packets will be
	  printed.
	* liboping: The library has been relicensed under the LGPL 2.1.

2010-06-13, Version 1.4.0:
	* noping: A new front-end to liboping, using the ncurses library, has
	  been added. The new command line application displays ping
	  statistics online and highlights aberrant round-trip times.

2009-12-20, Version 1.3.4:
	* src/liboping.c: When one file descriptor was in an error state, the
	  select(2) loop would run indefinitely. Error handling has been
	  improved so the loop ends gracefully now in this case.
	* src/liboping.c: Drop privileges before reading files if supported by
	  the system. This way files are opened using the user's original
	  privileges when using the “-f” option.
	* Net::Oping: An off-by-one error has been fixed in the Perl bindings.
	  Thanks to Fredrik Soderblom for his patch.

2009-09-29, Version 1.3.3:
	* oping: Disable the “-f” option if the real and effective user IDs
	  don't match. If that is the case the program is probably running
	  SetUID and should not read foreign files. Unfortunately, dropping
	  privileges before reading the file is not possible, because they are
	  required for opening raw sockets.

	  Reading from STDIN using “-f -” is still possible.

	  Thanks to Steve Kemp for reporting this issue as Debian bug #548684.

2009-07-27, Version 1.3.2:
	* src/oping.h: Remove `HAVE_*_H' macros for system headers. Those
	  macros should not be used in system wide installed header files.
	  Thanks to Sebastian for fixing this.

2009-07-20, Version 1.3.1

tests/main.c
#include <oping.h>
#include <stdbool.h>
#include <netinet/in.h>
#include <stdio.h>

#include <sys/types.h>
#include <sys/socket.h>
#include <ifaddrs.h>
#include <string.h>

#include <arpa/inet.h>

#define DATA "=TOTO="

#define FUNCTION_ERROR(FORMAT, ARGS...) printf(FORMAT, ## ARGS); return -1

int main (int argc, char const *argv[])
{
  int             i;
  pingobj_t       *ping;
  pingobj_iter_t  *ph;
  struct ifaddrs  *addrs, *addr;
  char            *ip;
  bool            found = false;
  int             enable_arp = 1;
  char            source_address[INET_ADDRSTRLEN + 1];
  double          latency;
  char            address[100];
  
  
  if( argc < 3 ) {
    printf("Usage: %s <interface> <destination>\n", argv[0]);
    exit(1);
  }
  
  ping = ping_construct();
  if( ping == NULL ) {
    FUNCTION_ERROR("ping_construct failed\n");
  }
  
  // find the ip address attached to the interface we wants to bind on
  // if more than one address are attached to the interface just
  // take the first one
  //
  if( getifaddrs(&addrs) < 0 ) {
    FUNCTION_ERROR("Unable to get interface addresses");
  }
  
  for(addr = addrs; addr != NULL; addr = addr->ifa_next) {
    // check if this is the interface we want
    if( (strcmp(addr->ifa_name, argv[1]) == 0) && (addr->ifa_addr->sa_family == AF_INET) ) {
      // convert addr to human form
      ip = inet_ntoa(((struct sockaddr_in *)addr->ifa_addr)->sin_addr);
      strncpy((char *) source_address, ip, INET_ADDRSTRLEN);
      printf("%s : %s\n", addr->ifa_name, ip);
      found = true;
      break;
    }
  }
  
  if( found == false ) {
    FUNCTION_ERROR("Unable to find address for interface <%s>", argv[1]);
  }
  
  if( ping_setopt(ping, PING_OPT_SOURCE, (void*) source_address) != 0 ) {
    FUNCTION_ERROR("Setting source address to %s on %s failed: %s\n", 
      source_address, argv[1], ping_get_error(ping));
  }
  
  if( ping_setopt(ping, PING_OPT_DEVICE, (void *) argv[1]) != 0 ) {
    FUNCTION_ERROR("Setting source interface to %s failed: %s\n", 
      argv[1], ping_get_error(ping));
  }
  
  // payload
  if( ping_setopt(ping, PING_OPT_DATA, (void *) DATA) != 0 ) {
    FUNCTION_ERROR("Setting payload failed: %s\n", ping_get_error(ping));
  }
  
  // add the hosts
  for(i= 2; i< argc; i++) {
    if( ping_host_add(ping, argv[i]) < 0 ) {
      FUNCTION_ERROR("Unable to add host '%s' : %s\n", argv[i], ping_get_error(ping));
    }
  }
  
  if( ping_setopt(ping, PING_OPT_ARP, (void *) &enable_arp) != 0 ) {
    FUNCTION_ERROR("Setting arp mode failed: %s\n", ping_get_error(ping));
  }
  
  for(i = 0; i< 4; i++) {
    size_t len;
    
    ping_send(ping);
    
    for( ph = ping_iterator_get(ping); ph != NULL; ph = ping_iterator_next(ph)) {
      len = sizeof(latency);
      ping_iterator_get_info(ph, PING_INFO_LATENCY, &latency, &len);
      len = sizeof(address);
      ping_iterator_get_info(ph, PING_INFO_ADDRESS, address, &len);
      
      if( latency != -1 ){
        printf("latency [%s]: %f\n", address, latency * 1000);
      }
      else {
        printf("latency [%s]: timeout\n", address);
      }
    }
  }
  
  return 0;
}



src/arp.c
/**
 * Object oriented C module to send ARP `echo's.
 * Copyright (C) 2010  Julien Ammous <j.ammous at gmail.com>
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or (at your
 * option) any later version.
 * 
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include "config.h"

#include <stdlib.h>

#if HAVE_STDINT_H
# include <stdint.h>
#endif
#if HAVE_INTTYPES_H
# include <inttypes.h>
#endif

#include "oping.h"
#include "oping_private.h"

/*
 * Private functions
 */
static int arp_ping_send(pingobj_t *pingobj, pinghost_t *ph) /* {{{ */
{
  uint8_t ethnull[ETH_ALEN];
  uint8_t ethall[ETH_ALEN];
  libnet_ptag_t status;

  /* sender's protocol address */
  void *spa;
  /* targer protocol address */
  void *tpa;

  /* TODO: Check if this is an IPv4 address. If not, return an appropriate
   * error. */

  memset (ethnull, 0x00, sizeof (ethnull));
  memset (ethall,  0xff, sizeof (ethall));

  spa = &((struct sockaddr_in *) pingobj->srcaddr)->sin_addr;
  tpa = &((struct sockaddr_in *) ph->addr)->sin_addr;

  status = libnet_build_arp
  (
      /*   hardware address format */ ARPHRD_ETHER,
      /*   protocol address format */ ETHERTYPE_IP,
      /*   hardware address length */ ETH_ALEN,
      /*   protocol address length */ IP_ALEN,
      /*        ARP operation type */ ARPOP_REQUEST,
      /* sender's hardware address */ pingobj->srcmac,
      /* sender's protocol address */ spa,
      /*   target hardware address */ ethnull,
      /*   targer protocol address */ tpa,
      /*                   payload */ (uint8_t *) ph->data,
      /*            payload length */ (uint32_t) strlen (ph->data),
      /*            libnet context */ pingobj->ln,
      /*    build new protocol tag */ 0
  );
  if (status == -1)
  {
    dprintf("libnet_build_arp: %s", libnet_geterror(pingobj->ln));
    return -1;
  }
  
  status = libnet_build_ethernet
  (
      /* destination ethernet address */ ethall,
      /*      source ethernet address */ pingobj->srcmac,
      /*    upper layer protocol type */ ETHERTYPE_ARP,
      /*                      payload */ NULL,
      /*               payload length */ 0,
      /*               libnet context */ pingobj->ln,
      /*       build new protocol tag */ 0
  );
  if (status == -1)
  {
    dprintf("libnet_build_ethernet: %s", libnet_geterror(pingobj->ln));
    return -1;
  }
  
  if (gettimeofday (ph->timer, NULL) == -1)
  {
    timerclear (ph->timer);
    libnet_clear_packet (pingobj->ln);
    return -1;
  }
  
  ph->latency = -1.0;
  
  /* send packet */
  if (libnet_write(pingobj->ln) == -1)
  {
    libnet_clear_packet(pingobj->ln);
    dprintf ("libnet_write: %s", libnet_geterror(pingobj->ln));
    return -1;
  }
  
  libnet_clear_packet(pingobj->ln);
  return 0;
} /* }}} int arp_ping_send */

/* wait for all the answers
 * return conditions:
 * - all answers received
 * - timeout reached
 */
static void arp_pings_recv(pingobj_t *pingobj) /* {{{ */
{
  const uint8_t               *data;
  struct pcap_pkthdr          pkthdr;
  pinghost_t                  *host;
  uint32_t                    ip;
  struct libnet_ethernet_hdr  *heth;
  struct libnet_arp_hdr       *harp;
  struct timeval              diff;
  struct timeval              timeout, nowtime;
  
  int                         break_loop;
  
  if( gettimeofday(&nowtime, NULL) == -1 ) {
    dprintf("gettimeofday: %s\n", strerror(errno));
    return;
  }
  
  /* Set up timeout */
  timeout.tv_sec = (time_t) pingobj->timeout;
  timeout.tv_usec = (suseconds_t) (1000000 * (pingobj->timeout - ((double) timeout.tv_sec)));
  
  while (1) {
    data = pcap_next(pingobj->pcap, &pkthdr);
    if( data == NULL ) {
      dprintf("pcap_next returned NULL !\n");
      continue;
    }
    
    heth = (void*) data;
    harp = (void*)((char*)heth + LIBNET_ETH_H);
    
    memcpy(&ip, (char*)harp + harp->ar_hln + LIBNET_ARP_H, 4);
    
    /* check if the arp reply comes from a known host */
    for(host = pingobj->head; host != NULL; host = host->next) {
      
      timersub(&pkthdr.ts, host->timer, &diff);
      
      if( (host->latency >= 0.0) || (host->timeout_reached == 1) ) {
        continue;
      }
      
      if( ((struct sockaddr_in *) host->addr)->sin_addr.s_addr == ip ) {
        /* we found a matching host, compute latency */
        host->latency  = ((double) diff.tv_usec) / 1000000.0;
        host->latency += ((double) diff.tv_sec);
        
        dprintf("received ARP REPLY for %s\n", inet_ntoa( *((struct in_addr *) &ip)) );
      }
      
      timersub(&nowtime, host->timer, &diff);
      
      dprintf("timeout state: %f > %f\n", (diff.tv_sec * 1000.0) + (diff.tv_usec / 1000.0), pingobj->timeout*1000);
      
      if( ((double)diff.tv_sec + (diff.tv_usec / 1000000)) > pingobj->timeout ) {
        host->timeout_reached = 1;
      }
    }
    
    /* check if there are some hosts left */
    break_loop = 1;
    for(host = pingobj->head; host != NULL; host = host->next) {
      if( (host->timeout_reached == 0) && (host->latency < 0.0) ) {
        break_loop = 0;
        break;
      }
    }
    
    if( break_loop )
      break;
  }
} /* }}} void arp_pings_recv */

/*
 * Semi-public functions
 * (can be called from liboping.c)
 */
int ping_construct_arp (pingobj_t *pingobj) /* {{{ */
{
  struct bpf_program arp_p;
  char mac_addr[20];
  char pcap_filter[256];
  char ebuf[PCAP_ERRBUF_SIZE];
  uint8_t *cp;

  if (pingobj == NULL)
    return (EINVAL);

  /* Initialize the pointers only used here. The "memset (0)" in
   * "ping_construct" may not initialize them to NULL. */
  pingobj->pcap = NULL;
  pingobj->ln = NULL;

  memset (ebuf, 0, sizeof (ebuf));
  
  /* libnet init */
  pingobj->ln = libnet_init(LIBNET_LINK_ADV, pingobj->device, ebuf);
  if(pingobj->ln == NULL)
  {
    dprintf("libnet_init: %s\n", ebuf);
    return -1;
  }
  
  /* get interface mac address */
  cp = (uint8_t *) libnet_get_hwaddr(pingobj->ln);
  if (cp == NULL)
  {
    dprintf("arping: libnet_get_hwaddr(): %s\n", libnet_geterror(pingobj->ln));
    return -1;
  }
  memcpy(pingobj->srcmac, cp, ETH_ALEN);
  
  snprintf (mac_addr, sizeof (mac_addr),
      "%02"PRIx8":%02"PRIx8":%02"PRIx8":%02"PRIx8":%02"PRIx8":%02"PRIx8,
      cp[0], cp[1], cp[2], cp[3], cp[4], cp[5]);
  mac_addr[sizeof (mac_addr) - 1] = 0;
  dprintf("ping_construct_arp: mac_addr = %s\n", mac_addr);
  
  /* pcap init */
  pingobj->pcap = pcap_open_live (pingobj->device, 100, 0, 10, ebuf);
  if (pingobj->pcap == NULL)
  {
    libnet_destroy (pingobj->ln);
    pingobj->ln = NULL;
    dprintf("pcap_open_live failed: %s\n", ebuf);
    return -1;
  }
  
  if (strlen(ebuf) > 0)
    dprintf("warning: %s\n", ebuf);
  
  snprintf (pcap_filter, sizeof (pcap_filter),
      "arp && ether dst %s && arp[7] = 0x02",
      mac_addr);
  dprintf("ping_construct_arp: pcap_filter = %s\n", pcap_filter);
  
  /* compile pcap filter */
  if( pcap_compile(pingobj->pcap, &arp_p, pcap_filter, 0, -1) == -1 ) {
      dprintf("pcap_compile(): %s\n", pcap_geterr(pingobj->pcap));
      return -1;
  }
  
  if( pcap_setfilter(pingobj->pcap, &arp_p) == -1 ) {
      dprintf("pcap_setfilter(): %s\n", pcap_geterr(pingobj->pcap));
      return -1;
  }
  
  return 0;
} /* }}} int ping_construct_arp */

void ping_destroy_arp (pingobj_t *obj) /* {{{ */
{
  if (obj == NULL)
    return;

  if (obj->pcap != NULL)
  {
    pcap_close (obj->pcap);
    obj->pcap = NULL;
  }

  if (obj->ln != NULL)
  {
    libnet_destroy (obj->ln);
    obj->ln = NULL;
  }
} /* }}} void ping_destroy_arp */

int ping_send_all_arp (pingobj_t *obj) /* {{{ */
{
  pinghost_t *host;

  if ((obj == NULL) || (obj->ln == NULL) || (obj->pcap == NULL))
    return (EINVAL);
  
  for (host = obj->head; host != NULL; host = host->next)
    if (arp_ping_send (obj, host) < 0)
      return (-1);
  
  return (0);
} /* }}} int ping_send_all_arp */

int ping_receive_all_arp (pingobj_t *obj) /* {{{ */
{
  if ((obj == NULL) || (obj->ln == NULL) || (obj->pcap == NULL))
    return (EINVAL);

  arp_pings_recv (obj);
  return 0;
} /* }}} int ping_receive_all_arp */

/* vim: set sw=2 sts=2 et fdm=marker : */


src/oping_private.h
/**
 * Object oriented C module to send ICMP and ICMPv6 `echo's.
 * Copyright (C) 2006-2010  Florian octo Forster
 * Copyright (C) 2010       Julien Ammous
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or (at your
 * option) any later version.
 * 
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 *
 * Authors:
 *   Florian octo Forster <octo at verplant.org>
 *   Julien Ammous <j.ammous at gmail.com>
 */

#ifndef OPING_PRIVATE_H
#define OPING_PRIVATE_H 1

#ifndef ETH_ALEN
#define ETH_ALEN 6
#endif

#ifndef IP_ALEN
#define IP_ALEN 4
#endif

#define PING_ERRMSG_LEN 256

#if WITH_DEBUG
# define dprintf(...) printf ("%s[%4i]: %-20s: ", __FILE__, __LINE__, __FUNCTION__); printf (__VA_ARGS__)
#else
# define dprintf(...) /**/
#endif

#ifdef HAVE_PCAP_H
#include <pcap.h>
#endif

#ifdef HAVE_LIBNET_H
#include <libnet.h>
#endif


struct pinghost
{
	/* username: name passed in by the user */
	char                    *username;
	/* hostname: name returned by the reverse lookup */
	char                    *hostname;
	struct sockaddr_storage *addr;
	socklen_t                addrlen;
	int                      addrfamily;
	int                      fd;
	int                      ident;
	int                      sequence;
	struct timeval          *timer;
	double                   latency;
	uint32_t                 dropped;
	int                      recv_ttl;
	char                    *data;
#ifdef ENABLE_ARP
  int                      timeout_reached;
#endif

	void                    *context;

	struct pinghost         *next;
};

struct pingobj
{
	double                   timeout;
	int                      ttl;
	int                      addrfamily;
	char                    *data;

	struct sockaddr         *srcaddr;
	socklen_t                srcaddrlen;

	char                    *device;

	char                     errmsg[PING_ERRMSG_LEN];

	pinghost_t              *head;

	_Bool                   use_arp;
#ifdef ENABLE_ARP
	/* arp */
	pcap_t                  *pcap;
	libnet_t                *ln;
	uint8_t                 srcmac[ETH_ALEN];
#endif
};

#ifdef ENABLE_ARP
int ping_construct_arp(pingobj_t *obj);
void ping_destroy_arp (pingobj_t *obj);
int ping_send_all_arp(pingobj_t *obj);
int ping_receive_all_arp(pingobj_t *obj);
#else
# define ping_construct_arp(obj) (-1)
# define ping_destroy_arp(obj) /**/
# define ping_send_all_arp(obj) (-1)
# define ping_receive_all_arp(obj) (-1)
#endif

#endif /* OPING_PRIVATE_H */


bindings/perl/Makefile.PL
use 5.006;
use ExtUtils::MakeMaker;
use Config (%Config);

my @OPING_PREFIX = (qw(/opt/oping /usr /usr/local));
my $OPING_PREFIX;

my $OPING_CPPFLAGS;
my $OPING_LDDLFLAGS;
my $OPING_LIBS;
my $OPING_DEPEND;

# TOP_BUILDDIR is set by liboping's build system, so Net::Oping can link with
# the yet uninstalled library.
my $TOP_BUILDDIR;
my $TARGET_LIBDIR;

# Parse custom command line arguments.
for (my $i = 0; $i < @ARGV; $i++)
{
	if ($ARGV[$i] =~ m#^OPING_PREFIX=(.*[^/])#)
	{
		unshift (@OPING_PREFIX, $1);
		splice (@ARGV, $i, 1);
		$i--;
	}
	elsif ($ARGV[$i] =~ m#^TOP_BUILDDIR=(.*[^/])#)
	{
		$TOP_BUILDDIR = $1;
		# TOP_BUILDDIR is passed from bindings/, but we're currently in
		# bindings/perl/. If it is a relative path, we need to add an
		# extra `../' in order to compensate for this.
		if ($TOP_BUILDDIR !~ m#^/#)
		{
			$TOP_BUILDDIR = "../$TOP_BUILDDIR";
		}
		splice (@ARGV, $i, 1);
		$i--;
	}
	elsif ($ARGV[$i] =~ m#^TARGET_LIBDIR=(.*[^/])#)
	{
		# Only save TARGET_LIBDIR if it's not a standard system library
		# directory, such as /usr/lib.
		if (!is_system_libdir ($1))
		{
			$TARGET_LIBDIR = $1;
		}
		splice (@ARGV, $i, 1);
		$i--;
	}
}

if (!$TOP_BUILDDIR)
{
	for (my $i = 0; $i < @OPING_PREFIX; $i++)
	{
		if (!-e $OPING_PREFIX[$i] . '/include/oping.h')
		{
			next;
		}

		$OPING_PREFIX = $OPING_PREFIX[$i];
		print "Found <oping.h> in $OPING_PREFIX/include\n";
		last;
	}
}

if ($TOP_BUILDDIR)
{
	# Use LDDLFLAGS here instead of LIBS, because:
	#  1) We need to make sure our library path comes first (and no locally
	#     installed version is used).
	#  2) Prevent MakeMaker from stipping the -rpath option when the
	#     library is to be installed in a non-standard path. Standard-paths
	#     are read from $Config{'libsdirs'} above.
	$OPING_CPPFLAGS = "-I$TOP_BUILDDIR/src";
	$OPING_LDDLFLAGS = "-L$TOP_BUILDDIR/src/.libs " . $Config{'lddlflags'};
	$OPING_LIBS = "-L$TOP_BUILDDIR/src/.libs -loping";

	if ($TARGET_LIBDIR)
	{
		$OPING_LDDLFLAGS .= qq( -Wl,-rpath -Wl,"$TARGET_LIBDIR");
	}

	$OPING_DEPEND = { 'Oping.o' => "$TOP_BUILDDIR/src/liboping.la" };
}
elsif ($OPING_PREFIX)
{
	# -rpath is automagically set in this case.
	$OPING_CPPFLAGS = "-I$OPING_PREFIX/include";
	$OPING_LIBS = "-L$OPING_PREFIX/lib -loping";
}

if (!$OPING_CPPFLAGS)
{
	my $search_path = join (', ', @OPING_PREFIX);
	print STDERR <<EOF;
******************************************************************************
* ERROR: COULD NOT FIND THE NEEDED HEADER FILE <oping.h>!                    *
******************************************************************************
The <oping.h> header file could not be found in the usual places. The prefix
paths searched right now are:
  $search_path

Please rerun Makefile.PL giving the prefix to the oping library using the
`OPING_PREFIX' argument. If you, for example, had installed the oping library
to /tmp/oping, the header file would be at /tmp/oping/include/oping.h; you'd
then need to run the Makefile.PL as follows:
  perl Makefile.PL OPING_PREFIX=/tmp/oping

As you can see, the argument passed via `OPING_PREFIX' must be the same
directory you passed to the liboping configure script using the `--prefix'
argument.

No Makefile has been created.
EOF
	exit (0);
}

WriteMakefile(
    NAME              => 'Net::Oping',
    VERSION_FROM      => 'lib/Net/Oping.pm',
    PREREQ_PM         => {},
    ($] >= 5.005
     ? (ABSTRACT_FROM => 'lib/Net/Oping.pm',
        AUTHOR        => 'Florian Forster <ff@octo.it>')
     : ()),
    ($OPING_DEPEND ? (depend => $OPING_DEPEND) : ()),
    LIBS              => [$OPING_LIBS],
    ($OPING_LDDLFLAGS ? (LDDLFLAGS => "$OPING_LDDLFLAGS") : ()),
    DEFINE            => '',
    INC               => "$OPING_CPPFLAGS"
);

sub is_system_libdir
{
	my $path = shift;
	for (split (' ', $Config{'libsdirs'}))
	{
		if ($path eq $_)
		{
			return (1);
		}
	}
	return;
}


src/oping.h
/**
 * Object oriented C module to send ICMP and ICMPv6 `echo's.
 * Copyright (C) 2006-2017  Florian octo Forster <ff at octo.it>
 *
 * This library is free software; you can redistribute it and/or modify it
 * under the terms of the GNU Lesser General Public License as published by the
 * Free Software Foundation; either version 2.1 of the License, or (at your
 * option) any later version.
 * 
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
 * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License
 * for more details.
 * 
 * You should have received a copy of the GNU Lesser General Public License
 * along with this library; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#ifndef OCTO_PING_H
#define OCTO_PING_H 1

#if HAVE_CONFIG_H
# include <config.h>
#endif

#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/socket.h>

#ifdef __cplusplus
extern "C" {
#endif

#define OPING_VERSION 1009000

/*
 * Type definitions
 */
struct pinghost;
typedef struct pinghost pinghost_t;

typedef pinghost_t pingobj_iter_t;

struct pingobj;
typedef struct pingobj pingobj_t;

#define PING_OPT_TIMEOUT 0x01
#define PING_OPT_TTL     0x02
#define PING_OPT_AF      0x04
#define PING_OPT_DATA    0x08
#define PING_OPT_SOURCE  0x10
#define PING_OPT_DEVICE  0x20
#define PING_OPT_QOS     0x40
#define PING_OPT_MARK    0x80

#define PING_DEF_TIMEOUT 1.0
#define PING_DEF_TTL     255
#define PING_DEF_AF      AF_UNSPEC
#define PING_DEF_DATA    "liboping -- ICMP ping library <http://octo.it/liboping/>"

/*
 * Method definitions
 */
pingobj_t *ping_construct (void);
void ping_destroy (pingobj_t *obj);

int ping_setopt (pingobj_t *obj, int option, void *value);

int ping_send (pingobj_t *obj);

int ping_host_add (pingobj_t *obj, const char *host);
int ping_host_remove (pingobj_t *obj, const char *host);

pingobj_iter_t *ping_iterator_get (pingobj_t *obj);
pingobj_iter_t *ping_iterator_next (pingobj_iter_t *iter);
int ping_iterator_count (pingobj_t *obj);

#define PING_INFO_HOSTNAME  1
#define PING_INFO_ADDRESS   2
#define PING_INFO_FAMILY    3
#define PING_INFO_LATENCY   4
#define PING_INFO_SEQUENCE  5
#define PING_INFO_IDENT     6
#define PING_INFO_DATA      7
#define PING_INFO_USERNAME  8
#define PING_INFO_DROPPED   9
#define PING_INFO_RECV_TTL 10
#define PING_INFO_RECV_QOS 11
int ping_iterator_get_info (pingobj_iter_t *iter, int info,
		void *buffer, size_t *buffer_len);

const char *ping_get_error (pingobj_t *obj);

void *ping_iterator_get_context (pingobj_iter_t *iter);
void  ping_iterator_set_context (pingobj_iter_t *iter, void *context);

#ifdef __cplusplus
}
#endif

#endif /* OCTO_PING_H */
